18 th July - NLP

In [57]:
import nltk
nltk.download('punkt')
nltk.download('averaged_perceptron_tagger')


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\sthaa\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\sthaa\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!


True

In [58]:
from textblob import TextBlob

text = "TextBlob is a great library for processing textual data."
blob = TextBlob(text)
print(blob.tags)


[('TextBlob', 'NNP'), ('is', 'VBZ'), ('a', 'DT'), ('great', 'JJ'), ('library', 'NN'), ('for', 'IN'), ('processing', 'VBG'), ('textual', 'JJ'), ('data', 'NNS')]


In [59]:
from textblob import TextBlob

In [60]:
myString = "John found a new coach and a new bed in his new apartment."

In [61]:
opt = TextBlob(myString)


In [62]:
#opt.tags

In [63]:
import nltk

In [64]:
# reg_exp = "NP: {<DT>?<JJ>*<NN>}"
reg_exp = "NP:{<DT>{1,3}<JJ>{1,3}<NN>{1,3}}"
rp = nltk.chunk.RegexpParser(reg_exp)

In [65]:
output = rp.parse(opt.tags)
print(output)

(S
  John/NNP
  found/VBD
  (NP a/DT new/JJ coach/NN)
  and/CC
  (NP a/DT new/JJ bed/NN)
  in/IN
  his/PRP$
  new/JJ
  apartment/NN)


In [66]:
#output.draw()

In [67]:
myString = "The little yellow dog barked at the cat."

In [68]:
reg_exp = r""" NP:
    {<.*>+}     # Chunk everything
    }<VBD|IN>+{ # Chink sequences of VBD nad IN
"""

In [69]:
rp = nltk.chunk.RegexpParser(reg_exp)
output = rp.parse(opt.tags)
print(output)

(S
  (NP John/NNP)
  found/VBD
  (NP a/DT new/JJ coach/NN and/CC a/DT new/JJ bed/NN)
  in/IN
  (NP his/PRP$ new/JJ apartment/NN))


In [70]:
# output.draw()

In [71]:
myString = "The little yellow dog barked at the cat and the cat escaped to the house."

In [72]:
reg_exp = r""" NP:
    {<.*>+}     # Chunk everything
    }<VBD|IN>+{ # Chink sequences of VBD nad IN
"""

In [73]:
rp = nltk.chunk.RegexpParser(reg_exp)
output = rp.parse(opt.tags)
print(output)

(S
  (NP John/NNP)
  found/VBD
  (NP a/DT new/JJ coach/NN and/CC a/DT new/JJ bed/NN)
  in/IN
  (NP his/PRP$ new/JJ apartment/NN))


In [74]:
reg_exp = r""" NP: 
{(<JJ>{1,2}<NN>{1,2})|(<DT>{1,2}<NN>{1,2})} # chunking 
}<VBD|IN>+{ # chinking
"""

In [75]:
rp = nltk.chunk.RegexpParser(reg_exp)
output = rp.parse(opt.tags)
print(output)


(S
  John/NNP
  found/VBD
  a/DT
  (NP new/JJ coach/NN)
  and/CC
  a/DT
  (NP new/JJ bed/NN)
  in/IN
  his/PRP$
  (NP new/JJ apartment/NN))


In [76]:
my_string = "John found a new coach and a new bed in his new apartment and the cat escaped to the house"
opt = TextBlob(my_string)
reg_exp = r""" NP: 
{(<JJ>{1,2}<NN>{1,2})|(<DT>{1,2}<NN>{1,2})} # chunking 
}<VBD|IN>+{ # chinking
"""
rp = nltk.chunk.RegexpParser(reg_exp)
output = rp.parse(opt.tags)
print(output)

(S
  John/NNP
  found/VBD
  a/DT
  (NP new/JJ coach/NN)
  and/CC
  a/DT
  (NP new/JJ bed/NN)
  in/IN
  his/PRP$
  (NP new/JJ apartment/NN)
  and/CC
  (NP the/DT cat/NN)
  escaped/VBD
  to/TO
  (NP the/DT house/NN))


# Module 8

In [77]:
from sklearn.datasets import fetch_20newsgroups

In [78]:
news = fetch_20newsgroups(subset="all")

In [79]:
print(len(news.data))

18846


In [80]:
import pandas as pd
data = {
    "news" : list(news.data),
    "target" : list(news.target)
}
df = pd.DataFrame(data)

In [81]:
df.head()

news  target
0  From: Mamatha Devineni Ratnam <mr47+@andrew.cm...      10
1  From: mblawson@midway.ecn.uoknor.edu (Matthew ...       3
2  From: hilmi-er@dsv.su.se (Hilmi Eren)\nSubject...      17
3  From: guyd@austin.ibm.com (Guy Dawson)\nSubjec...       3
4  From: Alexander Samuel McDiarmid <am2o+@andrew...       4

In [82]:
count_df = df.target.value_counts().reset_index()

Extract features from text files

In [83]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split

In [84]:
def train_test(classifier, X,y):
    X_train, X_test, y_train, y_test = train_test_split(X, y , test_size=0.25,random_state=48)
    classifier.fit(X_train,y_train)
    print("Train accuracy: {}".format(classifier.score(X_train, y_train)))
    print("Test accuracy: {}".format(classifier.score(X_test, y_test)))
    return classifier

In [85]:
from sklearn.naive_bayes import MultinomialNB

## Trail 1 - NB, Tfidf

In [86]:
trail1 = Pipeline([("vectorizer", TfidfVectorizer()),
                   ("classifier",MultinomialNB())])

In [87]:
train_test(trail1, news.data, news.target)

Train accuracy: 0.9254987972265459
Test accuracy: 0.8535653650254669


Pipeline(steps=[('vectorizer', TfidfVectorizer()),
                ('classifier', MultinomialNB())])

## Trail 2 - NB, Tfidf + stopwords

In [90]:
from nltk.corpus import stopwords
import nltk
nltk.download("stopwords")

[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\sthaa\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.


True

In [91]:
trail2 = Pipeline([("vectorizer", TfidfVectorizer(stop_words=stopwords.words("english"))),
                   ("classifier",MultinomialNB())])

In [92]:
train_test(trail2, news.data, news.target)

Train accuracy: 0.9472194707796802
Test accuracy: 0.8828522920203735


Pipeline(steps=[('vectorizer',
                 TfidfVectorizer(stop_words=['i', 'me', 'my', 'myself', 'we',
                                             'our', 'ours', 'ourselves', 'you',
                                             "you're", "you've", "you'll",
                                             "you'd", 'your', 'yours',
                                             'yourself', 'yourselves', 'he',
                                             'him', 'his', 'himself', 'she',
                                             "she's", 'her', 'hers', 'herself',
                                             'it', "it's", 'its', 'itself', ...])),
                ('classifier', MultinomialNB())])

## Trail 3 - NB + alpha, Tfidf + stopwords

In [93]:
trail3 = Pipeline([("vectorizer", TfidfVectorizer(stop_words=stopwords.words("english"))),
                   ("classifier",MultinomialNB(alpha=0.05))])

In [94]:
train_test(trail3, news.data, news.target)

Train accuracy: 0.9898118013301259
Test accuracy: 0.91553480475382


Pipeline(steps=[('vectorizer',
                 TfidfVectorizer(stop_words=['i', 'me', 'my', 'myself', 'we',
                                             'our', 'ours', 'ourselves', 'you',
                                             "you're", "you've", "you'll",
                                             "you'd", 'your', 'yours',
                                             'yourself', 'yourselves', 'he',
                                             'him', 'his', 'himself', 'she',
                                             "she's", 'her', 'hers', 'herself',
                                             'it', "it's", 'its', 'itself', ...])),
                ('classifier', MultinomialNB(alpha=0.05))])

## Trail 4 - SVM, Tfidf + stopwords + punctuation

In [95]:
import string
from sklearn import svm

In [96]:
trail4 = Pipeline([("vectorizer", TfidfVectorizer(stop_words=stopwords.words("english")+list(string.punctuation),min_df=5)),
                   ("classifier", svm.LinearSVC())])

In [97]:
train_test(trail4, news.data, news.target)

Train accuracy: 0.998584972406962
Test accuracy: 0.9276315789473685


Pipeline(steps=[('vectorizer',
                 TfidfVectorizer(min_df=5,
                                 stop_words=['i', 'me', 'my', 'myself', 'we',
                                             'our', 'ours', 'ourselves', 'you',
                                             "you're", "you've", "you'll",
                                             "you'd", 'your', 'yours',
                                             'yourself', 'yourselves', 'he',
                                             'him', 'his', 'himself', 'she',
                                             "she's", 'her', 'hers', 'herself',
                                             'it', "it's", 'its', 'itself', ...])),
                ('classifier', LinearSVC())])

## Trail 5 - XGBoost, trail 4 all - Tfidf

In [98]:
!pip install xgboost


[notice] A new release of pip is available: 24.1.2 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [99]:
from xgboost import XGBClassifier
import xgboost as xgb

In [100]:
trail5 = Pipeline([("vectorizer", TfidfVectorizer(stop_words=stopwords.words("english")+list(string.punctuation),min_df=5)),
                   ("classifier", xgb.XGBClassifier())])

In [101]:
train_test(trail5, news.data, news.target)

Train accuracy: 0.999929248620348
Test accuracy: 0.8520797962648556


Pipeline(steps=[('vectorizer',
                 TfidfVectorizer(min_df=5,
                                 stop_words=['i', 'me', 'my', 'myself', 'we',
                                             'our', 'ours', 'ourselves', 'you',
                                             "you're", "you've", "you'll",
                                             "you'd", 'your', 'yours',
                                             'yourself', 'yourselves', 'he',
                                             'him', 'his', 'himself', 'she',
                                             "she's", 'her', 'hers', 'herself',
                                             'it', "it's", 'its', 'itself', ...])),
                ('classifier',
                 XGBClassifier(base_score=None, booster=...
                               feature_types=None, gamma=None, grow_policy=None,
                               importance_type=None,
                               interaction_constraints=None, learning_rate=None,
                               max_bin=None, max_cat_threshold=None,
                               max_cat_to_onehot=None, max_delta_step=None,
                               max_depth=None, max_leaves=None,
                               min_child_weight=None, missing=nan,
                               monotone_constraints=None, multi_strategy=None,
                               n_estimators=None, n_jobs=None,
                               num_parallel_tree=None,
                               objective='multi:softprob', ...))])

## Trail 6 - Random Forest, trail 4 all - Tfidf

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
rf = RandomForestClassifier()

In [ ]:
trail6 = Pipeline([("vectorizer", TfidfVectorizer(stop_words=stopwords.words("english")+list(string.punctuation),min_df=5)),
                   ("classifier", rf)])

In [ ]:
train_test(trail6, news.data, news.target)

Train accuracy: 0.999929248620348
Test accuracy: 0.8514431239388794


Pipeline(steps=[('vectorizer',
                 TfidfVectorizer(min_df=5,
                                 stop_words=['i', 'me', 'my', 'myself', 'we',
                                             'our', 'ours', 'ourselves', 'you',
                                             "you're", "you've", "you'll",
                                             "you'd", 'your', 'yours',
                                             'yourself', 'yourselves', 'he',
                                             'him', 'his', 'himself', 'she',
                                             "she's", 'her', 'hers', 'herself',
                                             'it', "it's", 'its', 'itself', ...])),
                ('classifier', RandomForestClassifier())])

| Trial Number | Trial Details                             | Train Accuracy       | Test Accuracy        | Notes       |
|--------------|-------------------------------------------|----------------------|----------------------|-------------|
| 1            | MultinomialNB, vectorizer TFidf           | 0.9254987972265459   | 0.8535653650254669   | Normal      |
| 2            | NB, tfidf + stopwords                     | 0.9472194707796802   | 0.8828522920203735   | Normal      |
| 3            | NB + alpha, tfidf + stopwords             | 0.9898118013301259   | 0.91553480475382     | Normal      |
| 4            | SVM(classifier), tfidf + stopwords + punctuation | 0.998584972406962 | 0.9276315789473685   | Overfitted  |
| 5            | XGBoost, tfidf - trail 4 all              | 0.999929248620348    | 0.8520797962648556   | Overfitted  |
| 6            | Random forest, tfidf - trail 4 all        | 0.999929248620348    | 0.8514431239388794   | Overfitted  |
